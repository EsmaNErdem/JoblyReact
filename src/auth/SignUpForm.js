import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import Alert from "../utilities/Alert";
import "./SignUpForm.css"

/**
 * SignUpForm Component
 * Displays form and renders controlled components functionality
 * 
 * A form component for user registration. Handles user input for username, password, first name, last name, and email.
 * Displays form errors if registration fails.
 * On submission:
 * - calls signup function prop
 * - redirects to / route
 *
 * Routes ==> SignupForm ==> Alert
 * Routed as /signup
 * 
 */

const SignUpForm = ({ signup }) => {
    const history = useHistory();
    const INITIAL_STATE = {
        username: "",
        password: "",
        firstName: "",
        lastName: "",
        email: ""
    }
    const [formData, setFormData] = useState(INITIAL_STATE)
    const [formErrors, setFormErrors] = useState([]);

    console.debug(
        "SignupForm",
        "signup=", typeof signup,
        "formData=", formData,
        "formErrors=", formErrors,
    );

    /**
     * Handles form submission. Calls the signup function with the form data.
     * Redirects to the home page upon successful registration, or displays form errors.
     */
    const handleSubmit = async e => {
        e.preventDefault();
        const result =  await signup(formData)
        if(result.success){
            history.push("/")
        } else {
            setFormErrors(result.error)
        }
    }

    /** Update form data field */
    const handleChange = e => {
        const { name, value } = e.target
        setFormData(data => ({
            ...data,
            [name] : value
        }))
    }

    return (
        <div>
            <form onSubmit={handleSubmit}  className="SignUpForm">
                <label htmlFor="username">Username</label>
                <input
                    id="username"
                    name="username"
                    value={formData.username}
                    onChange={handleChange}
                    placeholder="Username"
                />
                <label htmlFor="password">Password</label>
                <input
                    id="password"
                    name="password"
                    type="password"
                    value={formData.password}
                    onChange={handleChange}
                    placeholder="Password"
                />
                <label htmlFor="firstName">First Name</label>
                <input
                    id="firstName"
                    name="firstName"
                    value={formData.firstName}
                    onChange={handleChange}
                    placeholder="First Name"
                />
                <label htmlFor="lastName">Last Name</label>
                <input
                    id="lastName"
                    name="lastName"
                    value={formData.lastName}
                    onChange={handleChange}
                    placeholder="Last Name"
                />
                <label htmlFor="email">Email</label>
                <input
                    id="email"
                    name="email"
                    type="email"
                    value={formData.email}
                    onChange={handleChange}
                    placeholder="email"
                />

                {formErrors.length
                    ? <Alert type="danger" messages={formErrors} />
                    : null
                }

                <button>Submit</button>
            </form>
        </div>
    )

}

export default SignUpForm