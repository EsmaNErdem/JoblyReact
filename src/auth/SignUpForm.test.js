import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';
import SignUpForm from "./SignUpForm";

test("renders without crashing", async () => {
    render(<SignUpForm />); 
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(<SignUpForm />);
    expect(asFragment()).toMatchSnapshot();
});

test("SignUpForm renders correctly", () => {
    const { getByLabelText, getByText } = render(<SignUpForm />);
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email");
    const submitButton = getByText("Submit");
  
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(firstNameInput).toBeInTheDocument();
    expect(lastNameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });
  
  test("Submitting the form calls signup function", () => {
    const mockSignUp = jest.fn().mockResolvedValue({ success: true }); // Mock login function
    const { getByLabelText, getByText } = render(
          <MemoryRouter >
              <SignUpForm signup={mockSignUp} />
          </MemoryRouter>
      );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const firstNameInput = getByLabelText("First Name");
    const lastNameInput = getByLabelText("Last Name");
    const emailInput = getByLabelText("Email");
      const submitButton = getByText("Submit");
  
    fireEvent.change(usernameInput, { target: { value: "username" } });
    fireEvent.change(passwordInput, { target: { value: 123456789 } });
    fireEvent.change(firstNameInput, { target: { value: "user" } });
    fireEvent.change(lastNameInput, { target: { value: "name" } });
    fireEvent.change(emailInput, { target: { value: "user@email.com" } });
    fireEvent.click(submitButton);
  
    expect(mockSignUp).toHaveBeenCalledWith({
      username: "username",
      password: "123456789",
      firstName: "user",
      lastName: "name",
      email: "user@email.com"
    });
  });
   
