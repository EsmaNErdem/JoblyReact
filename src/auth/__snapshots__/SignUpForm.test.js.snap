// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div>
    <form
      class="SignUpForm"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        id="username"
        name="username"
        placeholder="Username"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <label
        for="firstName"
      >
        First Name
      </label>
      <input
        id="firstName"
        name="firstName"
        placeholder="First Name"
        value=""
      />
      <label
        for="lastName"
      >
        Last Name
      </label>
      <input
        id="lastName"
        name="lastName"
        placeholder="Last Name"
        value=""
      />
      <label
        for="email"
      >
        Email
      </label>
      <input
        id="email"
        name="email"
        placeholder="email"
        type="email"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
